### PR TITLE
#2365 replace tree search

### LINKED ISSUE #2365: tree_search is too slow
Screen Shot 2022-10-13 at 5 15 22 PM

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

Fixes #2365 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by tinosulzer** [comment]
@abillscmu FYI

File: pybamm/discretisations/discretisation.py [hunk 1]
```diff
@@ -1214,43 +1214,42 @@ def check_variables(self, model):
                         )
                     )
 
-    def search_for_independent_var(self, model, var):
+    def search_for_independent_var(self, var, all_vars_in_eqns):
```
Comment (brosaplanella) [inline pybamm/discretisations/discretisation.py:1217]: I am a bit confused by the function names. Doesn't this one actually check if a given variable is independent, and the other one searches for independent variables in the model?

**Review by brosaplanella — APPROVED** [review]
Looks good, just see the minor comment

File: pybamm/discretisations/discretisation.py (hunk 1 above)
Comment (tinosulzer) [reply pybamm/discretisations/discretisation.py]: Are the new names better?

Comment (brosaplanella) [reply pybamm/discretisations/discretisation.py]: Yes, thanks!

**Pull request merged**